cabal-install/Distribution/Client/Dependency.hs

```diff
@@ -346,7 +348,7 @@ dontUpgradeNonUpgradeablePackages params =
   where
     extraConstraints =
       [ LabeledPackageConstraint
-        (PackageConstraintInstalled pkgname)
+        (PackageConstraint (unqualified pkgname) PackagePropertyInstalled)
```

Comment:
> So, if I understand you, I should modify PackageConstraint to support a wildcard or 'any' qualifier as an extra alternative, but leave the Qualifier and QPN datatypes as they are?

Yes, the solver still needs QPN to refer to a single instance of a package in many places.  It might be worth adding a new type for the scope of the constraint, since we'll probably need other scopes later, such as "all setup dependencies". For example:

```haskell
data PackageConstraint = PackageConstraint ConstraintScope PackageProperty

data ConstraintScope =
  Qualified QPN
  | AnyQualifier PackageName
  | AnySetupQualifier PackageName
```

> Let 'Unqualified' mean applies everywhere and add a new constructor Toplevel.

I think Toplevel is a lot clearer than the current use of Unqualified.  (I'm not suggesting a change for these PRs, though.)